tests/unit/test_scenario.py

```diff
@@ -2,33 +2,58 @@
 
 import ops
 import pytest
-from scenario import Context, Container, Relation, State
-
 from charm import HexanatorCharm
+from scenario import Container, Context, Relation, State
+
+
+def default_pebble_layer() -> dict:
+    import yaml
+
+    tmp = yaml.safe_load(open("rockcraft.yaml").read())
+    assert tmp["services"]["gubernator"]["startup"] == "enabled"
+    return tmp
 
-META = {
-    "name": "hexanator",
-    "requires": {
-        "ingress": {
-            "interface": "ingress"
-        }
-    },
-    "containers": {
-        "gubernator": {}
-    },
 
-}
+@pytest.fixture
+def ctx():
+    return Context(HexanatorCharm)
 
 
-def test_startup():
-    ctx = Context(HexanatorCharm, meta=META)
-    container = Container(name="gubernator", can_connect=True)
-    relation=Relation(endpoint="ingress", interface="ingress", remote_app_name="ingress")
-    state = State( leader=True, relations=[relation], containers=[container])
+@pytest.fixture
+def initial_state():
+    pebble_layers = {"default": ops.pebble.Layer(raw=default_pebble_layer())}  # type: ignore
+    container = Container("gubernator", can_connect=True, layers=pebble_layers)
+    ingress = Relation("ingress", id=0, interface="ingress", remote_app_name="ingress")
+    rate_limit = Relation("rate-limit", id=1, interface="http", remote_app_name="user")
+    return State(leader=True, relations={ingress, rate_limit}, containers={container})  # type: ignore
 
-    state = ctx.run(ctx.on.start(), state)
-    state = ctx.run(ctx.on.pebble_ready(container), state)
-    state = ctx.run(ctx.on.relation_joined(relation), state)
 
+def test_startup(ctx, initial_state):
+    container = list(initial_state.containers)[0]
+    state = ctx.run(ctx.on.pebble_ready(container), initial_state)
     assert state.unit_status == ops.ActiveStatus()
-    assert relation.local_app_data == {"model": ANY, "name": '"hexanator"', "port": "80", "strip-prefix": "true"}
+    assert (
+        list(state.containers)[0].service_status["gubernator"] == ops.pebble.ServiceStatus.ACTIVE
+    )
+
+
+def test_ingress(ctx, initial_state):
+    relation = initial_state.get_relation(0)
+    ctx.run(ctx.on.relation_joined(relation), initial_state)
+    assert relation.local_app_data == {
+        "model": ANY,
+        "name": '"hexanator"',
+        "port": "80",
+        "strip-prefix": "true",
+    }
+
+
+def test_rate_limit(ctx, initial_state, monkeypatch):
+    monkeypatch.setattr(
+        "socket.getfqdn", lambda: "hexanator-0.hexanator-endpoints.mymodel.svc.cluster.local"
+    )
+    relation = initial_state.get_relation(1)
```

Comment:
@tonyandrewmeyer It's a pity we still need (semi-)magic numbers here. Any way to avoid?

Reply:
Ah, I see @dimaqq opened an issue to discuss this [here](https://github.com/canonical/ops-scenario/issues/155).

Reply:
It's only there because the relation object is hidden away in a fixture.

You could use `get_relations since you know there's only one for the endpoint in this case, or you could adjust the fixtures, so that the relation is a fixture that both the initial state and the test use, so that you have the id available.